### ISSUE TITLE
Temporary columns are dropped too early

### ISSUE THREAD

**Issue by hombit** [lincc-frameworks/tape#286]
Sometimes temporary columns are dropped just after just being computed. This code shows a message about `nobs_total` being dropped and raises an error that it does not exist.
```python
from tape import Ensemble

ens = Ensemble(client=False).from_dataset('s82_rrlyrae', table_sync=False)
# This line is crucial, it makes columns introduced in future to be expired
ens.query('error < 1e9', table='source')
ens.calc_nobs(temporary=True, by_band=False)
ens.query('nobs_total > 100', table='object')
```

<details><summary>expand</summary>

```

/opt/homebrew/lib/python3.11/site-packages/dask/dataframe/core.py:5243: UserWarning: New index has same name as existing, this is a no-op.
  warnings.warn(
/opt/homebrew/lib/python3.11/site-packages/dask/dataframe/core.py:5243: UserWarning: New index has same name as existing, this is a no-op.
  warnings.warn(







Temporary columns dropped from Object Table: ['nobs_total']
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
File /opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/scope.py:231, in Scope.resolve(self, key, is_local)
    230 if self.has_resolvers:
--> 231     return self.resolvers[key]
    233 # if we're here that means that we have no locals and we also have
    234 # no resolvers

File /opt/homebrew/Cellar/python@3.11/3.11.6/Frameworks/Python.framework/Versions/3.11/lib/python3.11/collections/__init__.py:1006, in ChainMap.__getitem__(self, key)
   1005         pass
-> 1006 return self.__missing__(key)

File /opt/homebrew/Cellar/python@3.11/3.11.6/Frameworks/Python.framework/Versions/3.11/lib/python3.11/collections/__init__.py:998, in ChainMap.__missing__(self, key)
    997 def __missing__(self, key):
--> 998     raise KeyError(key)

KeyError: 'nobs_total'

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
File /opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/scope.py:242, in Scope.resolve(self, key, is_local)
    238 try:
    239     # last ditch effort we look in temporaries
    240     # these are created when parsing indexing expressions
    241     # e.g., df[df > 0]
--> 242     return self.temps[key]
    243 except KeyError as err:

KeyError: 'nobs_total'

The above exception was the direct cause of the following exception:

UndefinedVariableError                    Traceback (most recent call last)
File /opt/homebrew/lib/python3.11/site-packages/dask/dataframe/utils.py:194, in raise_on_meta_error(funcname, udf)
    193 try:
--> 194     yield
    195 except Exception as e:

File /opt/homebrew/lib/python3.11/site-packages/dask/dataframe/core.py:6886, in _emulate(func, udf, *args, **kwargs)
   6885 with raise_on_meta_error(funcname(func), udf=udf), check_numeric_only_deprecation():
-> 6886     return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))

File /opt/homebrew/lib/python3.11/site-packages/dask/utils.py:1110, in methodcaller.__call__(self, _methodcaller__obj, *args, **kwargs)
   1109 def __call__(self, __obj, *args, **kwargs):
-> 1110     return getattr(__obj, self.method)(*args, **kwargs)

File /opt/homebrew/lib/python3.11/site-packages/pandas/core/frame.py:4602, in DataFrame.query(self, expr, inplace, **kwargs)
   4601 kwargs["target"] = None
-> 4602 res = self.eval(expr, **kwargs)
   4604 try:

File /opt/homebrew/lib/python3.11/site-packages/pandas/core/frame.py:4728, in DataFrame.eval(self, expr, inplace, **kwargs)
   4726 kwargs["resolvers"] = tuple(kwargs.get("resolvers", ())) + resolvers
-> 4728 return _eval(expr, inplace=inplace, **kwargs)

File /opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/eval.py:336, in eval(expr, parser, engine, local_dict, global_dict, resolvers, level, target, inplace)
    328 env = ensure_scope(
    329     level + 1,
    330     global_dict=global_dict,
   (...)
    333     target=target,
    334 )
--> 336 parsed_expr = Expr(expr, engine=engine, parser=parser, env=env)
    338 if engine == "numexpr" and (
    339     is_extension_array_dtype(parsed_expr.terms.return_type)
    340     or getattr(parsed_expr.terms, "operand_types", None) is not None
   (...)
    344     )
    345 ):

File /opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py:808, in Expr.__init__(self, expr, engine, parser, env, level)
    807 self._visitor = PARSERS[parser](self.env, self.engine, self.parser)
--> 808 self.terms = self.parse()

File /opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py:827, in Expr.parse(self)
    824 """
    825 Parse an expression.
    826 """
--> 827 return self._visitor.visit(self.expr)

File /opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py:411, in BaseExprVisitor.visit(self, node, **kwargs)
    410 visitor = getattr(self, method)
--> 411 return visitor(node, **kwargs)

File /opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py:417, in BaseExprVisitor.visit_Module(self, node, **kwargs)
    416 expr = node.body[0]
--> 417 return self.visit(expr, **kwargs)

File /opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py:411, in BaseExprVisitor.visit(self, node, **kwargs)
    410 visitor = getattr(self, method)
--> 411 return visitor(node, **kwargs)

File /opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py:420, in BaseExprVisitor.visit_Expr(self, node, **kwargs)
    419 def visit_Expr(self, node, **kwargs):
--> 420     return self.visit(node.value, **kwargs)

File /opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py:411, in BaseExprVisitor.visit(self, node, **kwargs)
    410 visitor = getattr(self, method)
--> 411 return visitor(node, **kwargs)

File /opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py:718, in BaseExprVisitor.visit_Compare(self, node, **kwargs)
    717     binop = ast.BinOp(op=op, left=node.left, right=comps[0])
--> 718     return self.visit(binop)
    720 # recursive case: we have a chained comparison, a CMP b CMP c, etc.

File /opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py:411, in BaseExprVisitor.visit(self, node, **kwargs)
    410 visitor = getattr(self, method)
--> 411 return visitor(node, **kwargs)

File /opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py:531, in BaseExprVisitor.visit_BinOp(self, node, **kwargs)
    530 def visit_BinOp(self, node, **kwargs):
--> 531     op, op_class, left, right = self._maybe_transform_eq_ne(node)
    532     left, right = self._maybe_downcast_constants(left, right)

File /opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py:451, in BaseExprVisitor._maybe_transform_eq_ne(self, node, left, right)
    450 if left is None:
--> 451     left = self.visit(node.left, side="left")
    452 if right is None:

File /opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py:411, in BaseExprVisitor.visit(self, node, **kwargs)
    410 visitor = getattr(self, method)
--> 411 return visitor(node, **kwargs)

File /opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py:544, in BaseExprVisitor.visit_Name(self, node, **kwargs)
    543 def visit_Name(self, node, **kwargs):
--> 544     return self.term_type(node.id, self.env, **kwargs)

File /opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/ops.py:91, in Term.__init__(self, name, env, side, encoding)
     90 self.is_local = tname.startswith(LOCAL_TAG) or tname in DEFAULT_GLOBALS
---> 91 self._value = self._resolve_name()
     92 self.encoding = encoding

File /opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/ops.py:115, in Term._resolve_name(self)
    113     is_local = False
--> 115 res = self.env.resolve(local_name, is_local=is_local)
    116 self.update(res)

File /opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/scope.py:244, in Scope.resolve(self, key, is_local)
    243 except KeyError as err:
--> 244     raise UndefinedVariableError(key, is_local) from err

UndefinedVariableError: name 'nobs_total' is not defined

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
Cell In [1], line 7
      5 ens.query('error < 1e9', table='source')
      6 ens.calc_nobs(temporary=True, by_band=False)
----> 7 ens.query('nobs_total > 100', table='object')

File ~/projects/lincc-frameworks/tape/src/tape/ensemble.py:419, in Ensemble.query(self, expr, table)
    417 self._lazy_sync_tables(table)
    418 if table == "object":
--> 419     self._object = self._object.query(expr)
    420     self._object_dirty = True
    421 elif table == "source":

File /opt/homebrew/lib/python3.11/site-packages/dask/dataframe/core.py:5475, in DataFrame.query(self, expr, **kwargs)
   5405 def query(self, expr, **kwargs):
   5406     """Filter dataframe with complex expression
   5407
   5408     Blocked version of pd.DataFrame.query
   (...)
   5473     2  1  3    2
   5474     """
-> 5475     return self.map_partitions(M.query, expr, **kwargs)

File /opt/homebrew/lib/python3.11/site-packages/dask/dataframe/core.py:1046, in _Frame.map_partitions(self, func, *args, **kwargs)
    918 @insert_meta_param_description(pad=12)
    919 def map_partitions(self, func, *args, **kwargs):
    920     """Apply Python function on each DataFrame partition.
    921
    922     Note that the index and divisions are assumed to remain unchanged.
   (...)
   1044     None as the division.
   1045     """
-> 1046     return map_partitions(func, self, *args, **kwargs)

File /opt/homebrew/lib/python3.11/site-packages/dask/dataframe/core.py:6956, in map_partitions(func, meta, enforce_metadata, transform_divisions, align_dataframes, *args, **kwargs)
   6949         raise ValueError(
   6950             f"{e}. If you don't want the partitions to be aligned, and are "
   6951             "calling `map_partitions` directly, pass `align_dataframes=False`."
   6952         ) from e
   6954 dfs = [df for df in args if isinstance(df, _Frame)]
-> 6956 meta = _get_meta_map_partitions(args, dfs, func, kwargs, meta, parent_meta)
   6957 if all(isinstance(arg, Scalar) for arg in args):
   6958     layer = {
   6959         (name, 0): (
   6960             apply,
   (...)
   6964         )
   6965     }

File /opt/homebrew/lib/python3.11/site-packages/dask/dataframe/core.py:7067, in _get_meta_map_partitions(args, dfs, func, kwargs, meta, parent_meta)
   7063     parent_meta = dfs[0]._meta
   7064 if meta is no_default:
   7065     # Use non-normalized kwargs here, as we want the real values (not
   7066     # delayed values)
-> 7067     meta = _emulate(func, *args, udf=True, **kwargs)
   7068     meta_is_emulated = True
   7069 else:

File /opt/homebrew/lib/python3.11/site-packages/dask/dataframe/core.py:6885, in _emulate(func, udf, *args, **kwargs)
   6880 def _emulate(func, *args, udf=False, **kwargs):
   6881     """
   6882     Apply a function using args / kwargs. If arguments contain dd.DataFrame /
   6883     dd.Series, using internal cache (``_meta``) for calculation
   6884     """
-> 6885     with raise_on_meta_error(funcname(func), udf=udf), check_numeric_only_deprecation():
   6886         return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))

File /opt/homebrew/Cellar/python@3.11/3.11.6/Frameworks/Python.framework/Versions/3.11/lib/python3.11/contextlib.py:155, in _GeneratorContextManager.__exit__(self, typ, value, traceback)
    153     value = typ()
    154 try:
--> 155     self.gen.throw(typ, value, traceback)
    156 except StopIteration as exc:
    157     # Suppress StopIteration *unless* it's the same exception that
    158     # was passed to throw().  This prevents a StopIteration
    159     # raised inside the "with" statement from being suppressed.
    160     return exc is not value

File /opt/homebrew/lib/python3.11/site-packages/dask/dataframe/utils.py:215, in raise_on_meta_error(funcname, udf)
    206 msg += (
    207     "Original error is below:\n"
    208     "------------------------\n"
   (...)
    212     "{2}"
    213 )
    214 msg = msg.format(f" in `{funcname}`" if funcname else "", repr(e), tb)
--> 215 raise ValueError(msg) from e

ValueError: Metadata inference failed in `query`.

You have supplied a custom function and Dask is unable to
determine the type of output that that function returns.

To resolve this please provide a meta= keyword.
The docstring of the Dask function you ran should have more information.

Original error is below:
------------------------
UndefinedVariableError("name 'nobs_total' is not defined")

Traceback:
---------
  File "/opt/homebrew/lib/python3.11/site-packages/dask/dataframe/utils.py", line 194, in raise_on_meta_error
    yield
  File "/opt/homebrew/lib/python3.11/site-packages/dask/dataframe/core.py", line 6886, in _emulate
    return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/dask/utils.py", line 1110, in __call__
    return getattr(__obj, self.method)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pandas/core/frame.py", line 4602, in query
    res = self.eval(expr, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pandas/core/frame.py", line 4728, in eval
    return _eval(expr, inplace=inplace, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/eval.py", line 336, in eval
    parsed_expr = Expr(expr, engine=engine, parser=parser, env=env)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py", line 808, in __init__
    self.terms = self.parse()
                 ^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py", line 827, in parse
    return self._visitor.visit(self.expr)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py", line 411, in visit
    return visitor(node, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py", line 417, in visit_Module
    return self.visit(expr, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py", line 411, in visit
    return visitor(node, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py", line 420, in visit_Expr
    return self.visit(node.value, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py", line 411, in visit
    return visitor(node, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py", line 718, in visit_Compare
    return self.visit(binop)
           ^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py", line 411, in visit
    return visitor(node, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py", line 531, in visit_BinOp
    op, op_class, left, right = self._maybe_transform_eq_ne(node)
                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py", line 451, in _maybe_transform_eq_ne
    left = self.visit(node.left, side="left")
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py", line 411, in visit
    return visitor(node, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/expr.py", line 544, in visit_Name
    return self.term_type(node.id, self.env, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/ops.py", line 91, in __init__
    self._value = self._resolve_name()
                  ^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/ops.py", line 115, in _resolve_name
    res = self.env.resolve(local_name, is_local=is_local)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pandas/core/computation/scope.py", line 244, in resolve
    raise UndefinedVariableError(key, is_local) from err
```
</details>

**Comment by hombit** [comment]
I think what is happening here is that first `query` sets dirty flag and the second `query` makes things to sync. We need to change the order of these operations, by introducing different flavors of temporary columns or different logic for synchronization

**Comment by dougbrn** [comment]
I believe this is now fixed with #294 resolved. Due to #294, calc_nobs was not syncing before the generating the temporary columns. In the above example, I believe the first query then has the source_dirty state propagated through to the second query and the sync for the second query then removes the nobs_total column before the query executes (as you say). By having calc_nobs sync up front, this clears the source_dirty state before we generate the temporary column, and as calc_nobs doesn't set any dirty flags itself the sync state is clean for the second query.

I've tested the code block above with the latest version of TAPE and things seem to run, but @hombit please let me know if things are resolved on your end as well.

**Comment by hombit** [comment]
Thank you @dougbrn, everything works as expected now! I also checked it with `assign(temporary=True)` and it works perfectly
```python
from tape import Ensemble

ens = Ensemble(client=False).from_dataset('s82_rrlyrae', table_sync=False)
ens.query('error < 1e9', table='source')
amplitude = ens.batch(np.ptp, ens._flux_col, label='amplitude')
ens.assign(amplitude=amplitude['result'], temporary=True)
ens.query('amplitude > 0.1')
ens.object.compute()
```
